### PR TITLE
Ch247/implement description

### PR DESCRIPTION
**Pull request by nmanduj1**

### CODE REVIEW

File: api/queries.py
Comment (seerocode) [inline api/queries.py]: Is this a straggler comment?

File: api/schema.graphql
Comment (seerocode) [inline api/schema.graphql]: What does value mean again? Is this like 'non-binary' for category?

File: api/database.py
Comment (jnu) [inline api/database.py]: this will end up being a list of values, right? maybe the plural `values` would be a better name?

File: api/database.py
Comment (jnu) [inline api/database.py]: i wonder if this would be more intuitive as `category_value` or something? i worry that the name `value` might not seem logical in the context of `Target` -- my first impression of the "value" of a target might be the numerical goal instead of the category it's set for.

File: api/database.py
Comment (jnu) [inline api/database.py]: Same comment as for `Target.value` -- should we generally call this the `CategoryValue` and `category_value` instead of simply `Value`? maybe it's just me who's confused by this :)

File: api/database.py
Comment (jnu) [inline api/database.py]: not a big deal, but i thought that as long as the relationships between all these objects were defined, you didn't have to explicitly `add` them all to the session?

there's no problem here, my only concern is that this dummy data is starting to get kind of sprawling. i'm happy to brainstorm about how to clean it up, but not a big priority!

File: api/mutations.py
Comment (jnu) [inline api/mutations.py]: i believe you can save yourself the query on 105 by just setting the `value_id` here, is that right? it's been a while since i've done this in sql alchemy, but under the hood the database is only storing the `value_id` in the `entry` table anyway, so we shouldn't need to fetch the whole `Value` from the database.

File: api/mutations.py
Comment (jnu) [inline api/mutations.py]: just want to note that the TIMESTAMP type in sql can behave a little unpredictably when you pass in times. it's a bit better not to calculate the time in python and let the database itself compute it -- you can do this by passing in:

```py
from sqlalchemy.sql import func

session.query(Value).filter(Value.category_id == id).update({'deleted': func.now()})
```

the reason is that the DB and the Python code might be running in different timezones, and the DB itself doesn't store any timezone information with the TIMESTAMP type. so when you read that data back into Python you have to make a judgment call about what timezone it's referring to. it's best if we can assume everything in the db is in the db's timezone -- so here we should just tell the db to figure out what time it is.

File: api/queries.py
Comment (jnu) [inline api/queries.py]: [repeating this comment again, sorry!] Query.value seems like it'll be confusing, but i'll defer to you all!

**Review by jnu — COMMENTED** [review]
made a few comments! i'm happy to explain any of them in detail and/or pair on any other things you have in flight here. overall it looks very thorough and solid, great work!

File: api/mutations.py
Comment (nmanduj1) [reply api/mutations.py]: Great catch! I just learned something new!! I will revisit this where necessary- I'm thinking some of these can be removed entirely given the changes we've made. Thanks joe!!

File: api/database.py
Comment (nmanduj1) [reply api/database.py]: I will tinker with it- I am sure there are ways to clean this up. It just wasn't a priority, but you're right that we should keep our code tight!

File: api/queries.py
Comment (nmanduj1) [reply api/queries.py]: For sure! I was also debating- naming can be so hard!!  I updated this to `category_value`

File: api/mutations.py
Comment (nmanduj1) [reply api/mutations.py]: Thanks for this explanation, Joe! I will update this- time has always been hard for me to wrap my head around. I should spend some time reading into it

File: api/schema.graphql
Comment (nmanduj1) [reply api/schema.graphql]: I updated this to `category_value`, like we discussed yesterday 👍

File: api/queries.py
Comment (nmanduj1) [reply api/queries.py]: Yea, Its from my figuring out how to elegantly handle error messaging. sorry! I have a few comments laying around- I will clean them up before opening a formal PR. Thank you for checking out my draft!!

File: api/mutations.py
Comment (nmanduj1) [reply api/mutations.py]: Hmm, @jnu I updated this and it seems I also have to add a `synchronize_session ` flag to `.update()` in order for that `.update()` to run as expected. My inclination is to pass a `False` flag, given that we are (currently) always using a `.commit()` immediately after updating. In the event we are continuing to use the objects in-memory, we'd probably want that flag to be set to `fetch`-

File: api/mutations.py
Comment (jnu) [inline api/mutations.py]: i have three suggestions here:
1) the name cleaning logic is duplicated here and in the `Category` field validator you wrote (which btw is awesome! validators are great!). This means we have to keep the logic in sync in both places, so if we ever found a bug or wanted to change the implementation we'd have to remember to update both functions. It's better to keep this all in one single function so that it's simpler to maintain if we want to change the implementation. And speaking of changing the implementation ... :p
2) a subtle thing, the order of the string manipulation matters here:
```py
name.capitalize().strip() != name.strip().capitalize()
```
we want to strip _before_ we capitalize, otherwise if there's whitespace at the beginning of the name the capitalization will have no effect
3) lastly, as you saw in my PR, I'm a fan of adding helper methods to the database.py classes to encapsulate common query patterns. In this way you can consolidate all the logic about how to standardize / clean / validate names in the `Category` class itself. This is good object-oriented practice and simplifies maintaining the code in the long run.

I would refactor to do something like:
```py
class Category(Base):
    @classmethod
    def get_by_name(cls, session, name):
        return session.query(cls).filter(cls.name == cls.clean_name(name)).first()

    @classmethod
    def clean_name(cls, name):
        return name.capitalize().strip()

    @validates('name')
    def capitalize_category(self, key, name):
        # NOTE: `self.__class__` basically just means `Category` here. It's slightly
        # better to avoid referencing the class explicitly by name in case a) we change
        # the name of the class, or b) we extend the class and want to allow the child
        # to override the method.
        return self.__class__.clean_name(name)
```

then here you can call:

```py
existing_category = Category.get_by_name(session, input["name"])
```
and now the mutations don't need to know anything about how to standardize the `name` input, it's all handled by the `Category` class itself 😊

File: api/schema.graphql
Comment (jnu) [inline api/schema.graphql]: smol nit: we should be consistent about case here - generally this schema uses camelCase (e.g. on line 102 in Entry.categoryValue)

File: api/database.py
Comment (jnu) [inline api/database.py:369]: wow nice, this is so much fewer lines of code!! love it <3

File: api/database.py
Comment (jnu) [inline api/database.py:292]: just checking, you mentioned we're not filling this column out currently, correct? After we merge #67 we can track inputter with `info.context['current_user']`

**Review by jnu — APPROVED** [review]
this looks great nancy, nice work! i made a couple of suggestions that should be quick to address.

Unfortunately this is going to have some painful merge conflicts with #67, especially with the test updates. If you want to merge this this morning I'm happy to solve the conflicts on my end.

**Pull request merged**
